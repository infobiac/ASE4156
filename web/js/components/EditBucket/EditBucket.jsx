// @flow
import React from 'react';
import PropTypes from 'prop-types';
import Button from 'material-ui/Button';
import TextField from 'material-ui/TextField';
import Dialog, {
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
} from 'material-ui/Dialog';
import Checkbox from 'material-ui/Checkbox';
import { FormGroup, FormControlLabel } from 'material-ui/Form';

type Props = {
  cancel: () => void,
  save: (string, bool, number) => void,
  errors?: ?Array<Error>,
}
type State = {
  public: bool,
  bucketName: string,
  investment: number,
}

export default class EditBucket extends React.Component<Props, State> {
  static propTypes = {
    cancel: PropTypes.func.isRequired,
    save: PropTypes.func.isRequired,
  }
  constructor() {
    super();
    this.state = {
      public: false,
      bucketName: '',
      investment: 1000,
    };
  }
  clickCheckbox = () => this.setState(state => ({
    ...state,
    public: !state.public,
  }))
  bucketNameChange = (e: SyntheticInputEvent<>) => {
    const text = e.target.value;
    this.setState(() => ({
      bucketName: text,
    }));
  }
  investmentChange = (e: SyntheticInputEvent<>) => {
    const investment = parseFloat(e.target.value);
    this.setState(() => ({
      investment,
    }));
  }
  save = () => {
    this.props.save(this.state.bucketName, this.state.public, this.state.investment);
  }
  render() {
    return (
      <Dialog open onRequestClose={this.props.cancel}>
        <DialogTitle>Subscribe</DialogTitle>
        <DialogContent>
          <DialogContentText>
            Create a new risk bucket that people can invest in
          </DialogContentText>
          {
            this.props.errors ? this.props.errors.map(e => (
              <DialogContentText key={e.message}>{e.message}</DialogContentText>
            )) : null
          }
          <FormGroup row>
            <TextField
              autoFocus
              margin="dense"
              id="name"
              label="Bucket name"
              type="text"
              value={this.state.bucketName}
              onChange={this.bucketNameChange}
              fullWidth
            />
          </FormGroup>
          <FormGroup row>
            <TextField
              autoFocus
              margin="dense"
              id="investment"
              label="Investment"
              type="text"
              value={this.state.investment.toFixed(2)}
              onChange={this.investmentChange}
              fullWidth
            />
          </FormGroup>
          <FormGroup row>
            <FormControlLabel
              id="publicContainer"
              control={
                <Checkbox
                  id="public"
                  checked={this.state.public}
                  onChange={this.clickCheckbox}
                />
              }
              label="Public"
            />
          </FormGroup>
        </DialogContent>
        <DialogActions>
          <Button id="cancel" onClick={this.props.cancel} color="primary">
            Cancel
          </Button>
          <Button id="save" onClick={this.save} color="primary">
            Save
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}
