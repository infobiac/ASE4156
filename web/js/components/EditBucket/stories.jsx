// @flow
import React from 'react';
import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';
import EditBucket from './EditBucket';

storiesOf('EditBucketDontTest', module).add('Playground', () => (
  <EditBucket save={action('save')} cancel={action('cancel')} />
));
