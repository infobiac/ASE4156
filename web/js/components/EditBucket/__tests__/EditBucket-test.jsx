import React from 'react';
import { shallow } from 'enzyme';
import EditBucket from '../EditBucket';

describe('EditBucket', () => {
  it('exists', () => {
    expect(EditBucket).toBeTruthy();
  });

  it('renders', () => {
    const cancelFn = jest.fn();
    const saveFn = jest.fn();
    const editBucket = shallow(<EditBucket cancel={cancelFn} save={saveFn} />);
    expect(editBucket).toMatchSnapshot();
    expect(cancelFn.mock.calls.length).toBe(0);
    expect(saveFn.mock.calls.length).toBe(0);
  });

  it('cancels', () => {
    const cancelFn = jest.fn();
    const saveFn = jest.fn();
    const editBucket = shallow(<EditBucket cancel={cancelFn} save={saveFn} />);
    editBucket.find('#cancel').simulate('click');
    expect(cancelFn.mock.calls.length).toBe(1);
    expect(editBucket).toMatchSnapshot();
  });

  it('enter name', () => {
    const cancelFn = jest.fn();
    const saveFn = jest.fn();
    const editBucket = shallow(<EditBucket cancel={cancelFn} save={saveFn} />);
    editBucket.find('#name').simulate('change', { target: { value: 'bucketName' } });
    expect(editBucket.find('#name').props().value).toEqual('bucketName');
    expect(editBucket).toMatchSnapshot();
  });

  it('enter investment', () => {
    const cancelFn = jest.fn();
    const saveFn = jest.fn();
    const editBucket = shallow(<EditBucket cancel={cancelFn} save={saveFn} />);
    editBucket.find('#investment').simulate('change', { target: { value: '200' } });
    expect(editBucket.find('#investment').props().value).toEqual('200.00');
    expect(editBucket).toMatchSnapshot();
  });

  it('change public', () => {
    const cancelFn = jest.fn();
    const saveFn = jest.fn();
    const editBucket = shallow(<EditBucket cancel={cancelFn} save={saveFn} />);
    const cb = shallow(editBucket.find('#publicContainer').props().control);
    expect(cb).toMatchSnapshot();
    cb.simulate('change');
    expect(editBucket.state().public).toBe(true);
  });

  it('saves', () => {
    const cancelFn = jest.fn();
    const saveFn = jest.fn();
    const editBucket = shallow(<EditBucket cancel={cancelFn} save={saveFn} />);
    editBucket.find('#name').simulate('change', { target: { value: 'bucketName' } });
    editBucket.find('#investment').simulate('change', { target: { value: '200' } });
    editBucket.find('#save').simulate('click');
    expect(saveFn.mock.calls.length).toBe(1);
    expect(saveFn.mock.calls[0]).toEqual(['bucketName', false, 200.0]);
  });

  it('displays error', () => {
    const cancelFn = jest.fn();
    const saveFn = jest.fn();
    const editBucket = shallow(<EditBucket cancel={cancelFn} save={saveFn} errors={[Error('messed up')]} />);
    expect(editBucket).toMatchSnapshot();
  });
});
