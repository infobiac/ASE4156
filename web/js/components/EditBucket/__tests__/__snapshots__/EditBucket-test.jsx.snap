// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditBucket cancels 1`] = `
<withStyles(Dialog)
  onRequestClose={[Function]}
  open={true}
>
  <withStyles(DialogTitle)>
    Subscribe
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <withStyles(DialogContentText)>
      Create a new risk bucket that people can invest in
    </withStyles(DialogContentText)>
    <withStyles(FormGroup)
      row={true}
    >
      <TextField
        autoFocus={true}
        fullWidth={true}
        id="name"
        label="Bucket name"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="text"
        value=""
      />
    </withStyles(FormGroup)>
    <withStyles(FormGroup)
      row={true}
    >
      <TextField
        autoFocus={true}
        fullWidth={true}
        id="investment"
        label="Investment"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="text"
        value="1000.00"
      />
    </withStyles(FormGroup)>
    <withStyles(FormGroup)
      row={true}
    >
      <withStyles(FormControlLabel)
        control={
          <withStyles(Checkbox)
            checked={false}
            id="public"
            onChange={[Function]}
          />
        }
        id="publicContainer"
        label="Public"
      />
    </withStyles(FormGroup)>
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      color="primary"
      id="cancel"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      color="primary"
      id="save"
      onClick={[Function]}
    >
      Save
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;

exports[`EditBucket change public 1`] = `
<Checkbox
  checked={false}
  classes={
    Object {
      "checked": "MuiCheckbox-checked-2",
      "default": "MuiCheckbox-default-1",
      "disabled": "MuiCheckbox-disabled-3",
    }
  }
  id="public"
  indeterminate={false}
  indeterminateIcon={<pure(IndeterminateCheckBox) />}
  onChange={[Function]}
/>
`;

exports[`EditBucket displays error 1`] = `
<withStyles(Dialog)
  onRequestClose={[Function]}
  open={true}
>
  <withStyles(DialogTitle)>
    Subscribe
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <withStyles(DialogContentText)>
      Create a new risk bucket that people can invest in
    </withStyles(DialogContentText)>
    <withStyles(DialogContentText)
      key="messed up"
    >
      messed up
    </withStyles(DialogContentText)>
    <withStyles(FormGroup)
      row={true}
    >
      <TextField
        autoFocus={true}
        fullWidth={true}
        id="name"
        label="Bucket name"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="text"
        value=""
      />
    </withStyles(FormGroup)>
    <withStyles(FormGroup)
      row={true}
    >
      <TextField
        autoFocus={true}
        fullWidth={true}
        id="investment"
        label="Investment"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="text"
        value="1000.00"
      />
    </withStyles(FormGroup)>
    <withStyles(FormGroup)
      row={true}
    >
      <withStyles(FormControlLabel)
        control={
          <withStyles(Checkbox)
            checked={false}
            id="public"
            onChange={[Function]}
          />
        }
        id="publicContainer"
        label="Public"
      />
    </withStyles(FormGroup)>
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      color="primary"
      id="cancel"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      color="primary"
      id="save"
      onClick={[Function]}
    >
      Save
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;

exports[`EditBucket enter investment 1`] = `
<withStyles(Dialog)
  onRequestClose={[Function]}
  open={true}
>
  <withStyles(DialogTitle)>
    Subscribe
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <withStyles(DialogContentText)>
      Create a new risk bucket that people can invest in
    </withStyles(DialogContentText)>
    <withStyles(FormGroup)
      row={true}
    >
      <TextField
        autoFocus={true}
        fullWidth={true}
        id="name"
        label="Bucket name"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="text"
        value=""
      />
    </withStyles(FormGroup)>
    <withStyles(FormGroup)
      row={true}
    >
      <TextField
        autoFocus={true}
        fullWidth={true}
        id="investment"
        label="Investment"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="text"
        value="200.00"
      />
    </withStyles(FormGroup)>
    <withStyles(FormGroup)
      row={true}
    >
      <withStyles(FormControlLabel)
        control={
          <withStyles(Checkbox)
            checked={false}
            id="public"
            onChange={[Function]}
          />
        }
        id="publicContainer"
        label="Public"
      />
    </withStyles(FormGroup)>
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      color="primary"
      id="cancel"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      color="primary"
      id="save"
      onClick={[Function]}
    >
      Save
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;

exports[`EditBucket enter name 1`] = `
<withStyles(Dialog)
  onRequestClose={[Function]}
  open={true}
>
  <withStyles(DialogTitle)>
    Subscribe
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <withStyles(DialogContentText)>
      Create a new risk bucket that people can invest in
    </withStyles(DialogContentText)>
    <withStyles(FormGroup)
      row={true}
    >
      <TextField
        autoFocus={true}
        fullWidth={true}
        id="name"
        label="Bucket name"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="text"
        value="bucketName"
      />
    </withStyles(FormGroup)>
    <withStyles(FormGroup)
      row={true}
    >
      <TextField
        autoFocus={true}
        fullWidth={true}
        id="investment"
        label="Investment"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="text"
        value="1000.00"
      />
    </withStyles(FormGroup)>
    <withStyles(FormGroup)
      row={true}
    >
      <withStyles(FormControlLabel)
        control={
          <withStyles(Checkbox)
            checked={false}
            id="public"
            onChange={[Function]}
          />
        }
        id="publicContainer"
        label="Public"
      />
    </withStyles(FormGroup)>
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      color="primary"
      id="cancel"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      color="primary"
      id="save"
      onClick={[Function]}
    >
      Save
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;

exports[`EditBucket renders 1`] = `
<withStyles(Dialog)
  onRequestClose={[Function]}
  open={true}
>
  <withStyles(DialogTitle)>
    Subscribe
  </withStyles(DialogTitle)>
  <withStyles(DialogContent)>
    <withStyles(DialogContentText)>
      Create a new risk bucket that people can invest in
    </withStyles(DialogContentText)>
    <withStyles(FormGroup)
      row={true}
    >
      <TextField
        autoFocus={true}
        fullWidth={true}
        id="name"
        label="Bucket name"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="text"
        value=""
      />
    </withStyles(FormGroup)>
    <withStyles(FormGroup)
      row={true}
    >
      <TextField
        autoFocus={true}
        fullWidth={true}
        id="investment"
        label="Investment"
        margin="dense"
        onChange={[Function]}
        required={false}
        select={false}
        type="text"
        value="1000.00"
      />
    </withStyles(FormGroup)>
    <withStyles(FormGroup)
      row={true}
    >
      <withStyles(FormControlLabel)
        control={
          <withStyles(Checkbox)
            checked={false}
            id="public"
            onChange={[Function]}
          />
        }
        id="publicContainer"
        label="Public"
      />
    </withStyles(FormGroup)>
  </withStyles(DialogContent)>
  <withStyles(DialogActions)>
    <withStyles(Button)
      color="primary"
      id="cancel"
      onClick={[Function]}
    >
      Cancel
    </withStyles(Button)>
    <withStyles(Button)
      color="primary"
      id="save"
      onClick={[Function]}
    >
      Save
    </withStyles(Button)>
  </withStyles(DialogActions)>
</withStyles(Dialog)>
`;
